src/bsoncxx/types/value.cpp

```diff
@@ -40,15 +45,28 @@ value::value(const value& rhs) {
 }
 
 value& value::operator=(const value& rhs) {
-    _type = rhs._type;
+    if(_type && !rhs._type) {
```

Comment:
I think this block needs to run regardless of whether rhs has a type set.  If it has a type, we still want to clear out the union in lhs to be able to set it to something new.